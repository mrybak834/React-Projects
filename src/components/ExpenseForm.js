import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

export default class ExpenseForm extends React.Component {

    constructor(props){
        super(props);

        this.state = {
            description: '',
            note: '',
            ...props.expense,
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            timestamp: props.expense ? moment(props.expense.timestamp) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;

        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;

        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        // Accept no amount or a valid value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onCalendarDateChange = (timestamp) => {
        timestamp && this.setState(() => ({ timestamp }));
    };

    onCalendarFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();

        // Handle input errors
        this.setState(() => ({ error: '' }));
        if (!this.state.description.trim()){
            return this.setState(() => ({ error: 'Please enter a description!'}));
        }
        if (!this.state.amount){
            return this.setState(() => ({ error: 'Please enter an amount!'}));
        }

        this.props.onSubmit({
            description: this.state.description,
            note: this.state.note,
            amount: parseFloat(this.state.amount, 10) * 100,
            timestamp: this.state.timestamp.valueOf()
        });
    };

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input 
                        id="inputDescription"
                        type="text" 
                        placeholder="Description" 
                        autoFocus 
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        id="inputAmount"
                        type="text" 
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.timestamp}
                        onDateChange={this.onCalendarDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                    />
                    <textarea 
                        placeholder="Add a note for your expense (Optional)" 
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button >Add Expense</button>
                </form>
                {this.state.error && <p>{this.state.error}</p>}
            </div>
        )
    }
}