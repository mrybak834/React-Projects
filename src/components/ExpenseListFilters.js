import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, setSortByFilter, setStartDateFilter, setEndDateFilter } from '../redux/actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    
    state = {
        calendarFocused: null
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSelectChange = (e) => {
        this.props.setSortByFilter(e.target.value);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDateFilter(startDate);
        this.props.setEndDateFilter(endDate);
    };

    render () {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                />
                
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSelectChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId="start_date"
                    endDate={this.props.filters.endDate}
                    endDateId="end_date"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={2}
                    isOutsideRange={() => false}
                />
        
            </div>
        );
    }
}


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    setTextFilter: value => dispatch(setTextFilter(value)),
    setSortByFilter: value => dispatch(setSortByFilter(value)),
    setStartDateFilter: startDate => dispatch(setStartDateFilter(startDate)),
    setEndDateFilter: endDate => dispatch(setEndDateFilter(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);