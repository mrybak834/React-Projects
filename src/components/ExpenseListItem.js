import React from 'react';
import { Link } from 'react-router-dom'

const ExpenseListItem =  ({ id, description, amount, timestamp }) => (
    <div>
        <h3><Link to={`edit/${id}`}>{description}</Link></h3>
        <p>${amount} - Created At: {timestamp}</p>
    </div>
);

export default ExpenseListItem;