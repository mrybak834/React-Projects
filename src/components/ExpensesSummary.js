import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import getFilteredExpenses from '../redux/selectors/expenses';
import sumExpenses from '../redux/selectors/expenses-total';

export const ExpensesSummary = ({text}) => (
    <div className="page-header">
        <div className="content-container">
            {text}
            <div className="page-header__actions">
                <Link className="button" to="/new">Add Expense</Link>
            </div>
        </div>
    </div>
);

const mapStateToProps = (state) => ({
    text: sumExpenses(getFilteredExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpensesSummary);