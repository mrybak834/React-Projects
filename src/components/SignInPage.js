import React from 'react';
import { connect } from 'react-redux';
import { startSignIn } from '../redux/actions/auth';

export class SignInPage extends React.Component {
    render() {
        return (
            <div className="box-layout">
                <div className="box-layout__box">
                    <h1 className="box-layout__title">Expensify</h1>
                    <p>Easily manage your budget</p>
                    <button className="button" onClick={this.props.startSignIn}>Sign In with Google</button>
                </div>
            </div>
        );
    }
}


// const mapStateToProps = (state, props) => {
//     return {

//     };
// }

const mapDispatchToProps = (dispatch) => ({
    startSignIn: () => dispatch(startSignIn())
});

export default connect(undefined, mapDispatchToProps)(SignInPage);