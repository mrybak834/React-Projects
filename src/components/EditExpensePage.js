import React from 'react';
import ExpenseForm from '../components/ExpenseForm';
import { connect } from 'react-redux';
import { startEditExpense } from '../redux/actions/expenses';
import { startRemoveExpense } from '../redux/actions/expenses';

export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);

        this.props.history.push('/');
    };

    onClick = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    };

    render() {
        return (

            this.props.expense ? (
                <div>
                    <div className="page-header">
                        <div className="content-container">
                            <h1 className="page-header__title">Edit Expense: {this.props.expense.description}</h1>
                        </div>
                    </div>
                    <div className="content-container">
                        <ExpenseForm
                            expense={this.props.expense}
                            onSubmit={this.onSubmit}
                        />
                        <button
                            className="button button--secondary"
                            onClick={this.onClick}
                        >Remove</button>
                    </div>
                </div>
            ) : (
                    <div className="page-header">
                        <div className="content-container">
                            <h1 className="page-header__title">Invalid Expense</h1>
                        </div>
                    </div>
                )
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);