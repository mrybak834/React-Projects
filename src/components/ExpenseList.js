import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../redux/selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ) : (
                props.expenses.map((expense, index) => 
                    <ExpenseListItem 
                        key={index}
                        {...expense}/>
                )
            )
        }
    </div>
);


const mapStateToProps = (state) => ({
        expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);