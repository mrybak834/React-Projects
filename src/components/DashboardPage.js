import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';


const DashboardPage = () => (
    <div>
        <p>Dashboard Page</p>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default DashboardPage;