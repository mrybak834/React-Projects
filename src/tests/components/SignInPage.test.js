import React from 'react';
import { shallow } from 'enzyme';
import { SignInPage } from '../../components/SignInPage';

let startSignIn, wrapper;
beforeEach(() => {
    startSignIn = jest.fn();
    wrapper = shallow(<SignInPage startSignIn={startSignIn} />);
});

test('Component Snapshot', () => {

    expect(wrapper).toMatchSnapshot();
})

test('Sign In', () => {
    wrapper.find('button').simulate('click');

    expect(startSignIn).toHaveBeenCalled();
});