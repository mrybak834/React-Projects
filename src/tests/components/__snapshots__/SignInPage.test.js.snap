// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Snapshot 1`] = `
<div
  className="box-layout"
>
  <div
    className="box-layout__box"
  >
    <h1
      className="box-layout__title"
    >
      Expensify
    </h1>
    <p>
      Easily manager your budget
    </p>
    <button
      className="button"
      onClick={[MockFunction]}
    >
      Sign In with Google
    </button>
  </div>
</div>
`;
